### ISSUE TITLE
RFC Consider unvendoring components from the mono-repo

### ISSUE THREAD

**Issue by rth** [pyodide/pyodide#2014]
Currently, we have a mono-repo which makes keeping a consistent state between elements easier, but also has the following disadvantages,
 - we have to wait for a Pyodide release for any bug fixes/improvements in individual components (e.g. micropip) or when new packages are added
 - the amount of code might be a bit overwhelming to new contributors. More importantly, they need to have Pyodide built to do anything (or rely on the CI). We can improve the build situation somewhat, but in the end, anything that requires a working compiler will have some contribution threshold. On contrary, there are much more users willing to contribute to a pure Python or JavaScript package (that might load some WASM).
 - easier to re-use these components. In the context of both,
   - the core CPython working on the WASM  https://github.com/pyodide/pyodide/issues/2000 and https://github.com/ethanhs/python-wasm 
   - and the JupyterLite team [hacking micropip](https://github.com/jupyterlite/jupyterlite/blob/b38fc93fa5f5bc855a61a74e0188b3026a9ca286/packages/pyolite-kernel/py/piplite/piplite/piplite.py)
   
   IMO we would also see more contribution from those communities if some of the components are a bit more independent from the Pyodide monorepo.
  - our CI setup is getting quite complex, particularly if we want to auto-deploy all of the components on releases
 

So we could consider unvendoring the following components as separate packages under the Pyodide Github org. Though this would mean we need to load them somehow in the Pyodide monorepo when needed, and it might make local builds slightly more complex.

## micropip

Is a (critical) plugin for Pyodide and could be made a separate package, same as say https://github.com/ryanking13/pyodide-importer. Its tests/CI could pre-load and cache the dev Pyodide release. It would be a good illustration of how to write CI for a Pyodide dependent package.

Proposed versioning: semantic versioning, independent from Pyodide. Though Pyodide pins some micropip version for the release. 

## pyodide-packages

It could be a possibility.

Significant problems right now are that,
 - a) we build all packages for each commit #2013 
 - b) if there is a new package, people need to wait for the next release to get it reliably.
 
If packages folder were in a separate repo we could,
 - if a new package is added, deploy only the package added with the latest Pyodide release immediately
 - then for a new Pyodide release re-deploy them under a new URL as we are doing now.

while in the main Pyodide repo we would run tests only on some subset of these packages.

Though it's not very clear how to rebuild packages in this context, we a) we should never overwrite a previously deployed stable package b) it won't have an impact anyway due to caching. So from an implementation point, it's still not very clear.

Proposed versioning: would follow Pyodide releases.

## pyodide-build

Is a pure Python package, and assuming we implement a better CLI https://github.com/pyodide/pyodide/issues/1977 there could be use cases for it, independent from the compilation tasks. Also, it's doesn't need to be updated very often, maybe except for skipping some compilation flags or working on new features.  

There are lots of usability improvements that could be done there and would be easier if it wouldn't trigger the full build. Though indeed maybe testing it without the main pyodide repo is not too reliable.

## Other comments
 
 - maybe the JS package https://github.com/pyodide/pyodide/tree/main/src/js but the problem is it's tight coupling with pyproxy.gen.js
 - for the type conversion code and Pyodide python package those it makes sense to keep where they are IMO.
 - the downside of this proposal is that it would make synchronization between these components a bit harder. Also it would likely mean that some would require an independent documentation deployment, which is more work.

**Comment by wlach** [comment]
A middle path might be to keep the monorepo but make it easier to hack on (and distribute) the individual components in isolation?

**Comment by rth** [comment]
> A middle path might be to keep the monorepo but make it easier to hack on (and distribute) the individual components in isolation?

True, but if there is no CI testing those components independently it's difficult to be sure they are as independent as they should be.

**Comment by bollwyvl** [comment]
Yep, it's a lot to get to the sophistication to handle some of these concerns at scale/in an automated fashion. 

> we should never overwrite a previously deployed stable package 

Right. When a new upstream of the baseline pyodide infrastructure is (about to be) released, it should indeed trigger a rebuild of every package, and almost as importantly, the single (or predictable number of countable) file(s) index of _where_ each release is needs to be updated... having this checked in _as data_ someplace, is super handy.

> no CI testing those components independently

Distributing the testing, and increasing the sophistication of, to be executed during the the `test` stanzas in each `meta.yaml` might help this. And the de-coupled `conda build --no-test <recipe>` and `conda build --test <tarball>` pattern. Getting peoples' testing stacks up and running can be _very_ hard, of course! 

Though it has been discussed a few times, both here and over on `conda-forge`, what makes conda-forge tick (ha) is the automation and delegation. Even if figuring out how to [bootstrap an architecture](https://github.com/conda-forge/conda-forge.github.io/issues/1401) is out of scope, the architecture of the thing is relevant as something that might _eventually_ be needed.

There are a couple layers of tooling beyond just `conda` and `conda-build`:

- all feedstocks start from the [staged-recipes](https://github.com/conda-forge/staged-recipes/) "sparse monorepo"
  - this allows for a one-time automated/manual review of "yeah, we can support having that in the ecosystem" 
  - a single PR might contain one or more related recipes
    - and each recipe might create one or more packages, derived from the upstream
  - after merge, each recipe folder is _removed_ from this repo, and rendered into a full, CI-enabled repository with... 
-  [conda-smithy](https://github.com/conda-forge/conda-smithy) to keep per-package (or split packages from a single tarball in some cases) to update the state of a repo's local/CI tools based on about three yaml files
  - it in turn uses 
    - [conda-forge-pinning](https://github.com/conda-forge/conda-forge-pinning-feedstock) as the source of truth for e.g. what matrix of versions of compilers should be used 
    - [county-fair](https://github.com/regro/cf-graph-countyfair)
    - [rever](https://github.com/regro/rever)
- all of the feedstocks are maintained as submodules in a single [repo](https://github.com/conda-forge/feedstocks) 
  - it doesn't contain much tooling of its own... north of 10k folders, the structure is only useful with very careful CLI actions, and breaks most web/gui tools, so it's really more forensic
- a bunch of [bots](https://github.com/regro/cf-scripts/) keep updating these things, often requiring batching, etc.
  - these primary handle making PRs for:
    - upstream updates
    - migrations which change many feedstocks to e.g. start using new data   
- the actual packages, in addition to the primary CDN-backed distribution point, are also available on the [releases](https://github.com/conda-forge/releases/releases/) repo
- finally, we use [constructor](https://github.com/conda/constructor) to build variants like [miniforge](https://github.com/conda-forge/miniforge) (and its sister mambaforge)

**Comment by hoodmane** [comment]
I think `_pyodide/console.py` would be a good candidate to move into a separate Python package or even to upstream into Python stdlib. It would be nice also to add `@redirect_stdin` to `contextlib`.

**Comment by ryanking13** [comment]
We've unvendored micropip and pyodide-build. Now, the only remaining one is the recipes. I will update the progress in #4918 instead of this issue. Closing as resolved.
